resolver-dns/src/main/java/io/netty/resolver/dns/DefaultAuthoritativeDnsServerCache.java

```diff
@@ -0,0 +1,319 @@
+        e.scheduleExpiration(loop, new Runnable() {
+            @Override
+            public void run() {
+                // We always remove all entries for a hostname once one entry expire. This is not the
+                // most efficient to do but this way we can guarantee that if a DnsResolver
+                // be configured to prefer one ip family over the other we will not return unexpected
+                // results to the enduser if one of the A or AAAA records has different TTL settings.
+                //
+                // As a TTL is just a hint of the maximum time a cache is allowed to cache stuff it's
+                // completely fine to remove the entry even if the TTL is not reached yet.
+                //
+                // See https://github.com/netty/netty/issues/7329
+                Entries entries = resolveCache.remove(hostname);
+                if (entries != null) {
+                    entries.clearAndCancel();
+                }
+            }
+        }, ttl, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public String toString() {
+        return new StringBuilder()
+                .append("DefaultAuthoritativeDnsServerCache(minTtl=")
+                .append(minTtl).append(", maxTtl=")
+                .append(maxTtl).append(", cached nameservers=")
+                .append(resolveCache.size()).append(")")
+                .toString();
+    }
+
+    private static final class Entry {
+
+        volatile ScheduledFuture<?> expirationFuture;
+
+        private final InetSocketAddress address;
+
+        Entry(InetSocketAddress address) {
+            this.address = checkNotNull(address, "address");
+        }
+
+        InetSocketAddress nameServer() {
+            return address;
+        }
+
+        void scheduleExpiration(EventLoop loop, Runnable task, long delay, TimeUnit unit) {
+            assert expirationFuture == null : "expiration task scheduled already";
+            expirationFuture = loop.schedule(task, delay, unit);
+        }
+
+        void cancelExpiration() {
+            ScheduledFuture<?> expirationFuture = this.expirationFuture;
+            if (expirationFuture != null) {
+                expirationFuture.cancel(false);
+            }
+        }
+
+        @Override
+        public String toString() {
+            return address.toString();
+        }
+    }
+
+    // This List implementation always has RandomAccess semantics as we either wrap an ArrayList, empty List or
+    // one which has only one element.
+    private static final class EntryList extends AbstractList<InetSocketAddress> implements RandomAccess {
+
+        static final EntryList EMPTY = new EntryList(Collections.<Entry>emptyList());
+
+        private final List<Entry> entries;
+
+        EntryList(Entry entry) {
+            this(Collections.singletonList(entry));
+        }
+
+        EntryList(List<Entry> entries) {
+            this.entries = entries;
```

Comment:
Optional: make a defensive copy.

Reply:
I think thats not needed with the current implementation.